levelmap.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2021 Siôn le Roux.  All rights reserved.
+// Use of this source code is subject to an MIT-style
+// licence which can be found in the LICENSE file.
+
+package main
+
+import (
+	"image"
+	"math"
+
+	"github.com/fzipp/astar"
+)
+
+type LevelMap [][]int
+
+
+func CreateMap(w, h int) LevelMap {
+	lmap := make([][]int, h)
+	for i := range lmap {
+		lmap[i] = make([]int, w)
+	}
+	
+	return lmap
+}
+
+// Neighbours implements the astar.Graph interface
+func (m LevelMap) Neighbours(p image.Point) []image.Point {
+	offsets := []image.Point{
+		image.Pt(0, -1), // North
+		image.Pt(1, 0),  // East
+		image.Pt(0, 1),  // South
+		image.Pt(-1, 0), // West
+	}
+	offsetsDiag := []image.Point{
+		image.Pt(1, -1),  // NorthEast
+		image.Pt(1, 1),   // SouthEast
+		image.Pt(-1, 1),  // SouthWest
+		image.Pt(-1, -1), // NorthWest
+	}
+	res := make([]image.Point, 0, 8)
+	
+	// Check avaialable diagonal neighbours (free only if there is a wide enough corridor)
+	for _, off := range offsetsDiag {
+		q := p.Add(off)
+		q1 := p.Add(image.Pt(0, off.Y))
+		q2 := p.Add(image.Pt(off.X, 0))
+		if m.isFreeAt(q) && m.isFreeAt(q1) && m.isFreeAt(q2) {
+			res = append(res, q)
+		}
+	}
+
+	// Check avaialable  neighbours
+	for _, off := range offsets {
+		q := p.Add(off)
+		if m.isFreeAt(q) {
+			res = append(res, q)
+		}
+	}
+	return res
+}
+
+// isFreeAt returns if the cell is free
+func (m LevelMap) isFreeAt(p image.Point) bool {
+	return m[p.Y][p.X] == 0
+}
+
+// distance calculates Euclidean distance between the points
+func distance(p, q image.Point) float64 {
+	d := q.Sub(p)
+	return math.Sqrt(float64(d.X*d.X + d.Y*d.Y))
+}
+
+// SetObstacle sets the cell as obstacle
+func (m LevelMap) SetObstacle(x, y int) {
+	m[y][x] = 1
+}
+
+// FindPath finds path between two coordinates on map
+func (m LevelMap) FindPath(start, dest Coord) []Coord {
+	var result []Coord
+
+	startCell := image.Pt(int(start.X / 32), int(start.Y / 32))
+	destCell := image.Pt(int(dest.X / 32), int(dest.Y / 32))
+	apath := astar.FindPath[image.Point](m, startCell, destCell, distance, distance)
+	apath = simplifyPath(apath)
+	for _, p := range apath {
+		result = append(result, Coord{X: float64(p.X)*32+16, Y: float64(p.Y)*32+16})
```

Comment:
I'm amazed that you managed to mix image.Point and float64 in the same file and hardly had to do any manual type casts. Not sure if lucky or very clever. Btw these hard-coded `32` and `16` values are probably referring to some size thing that's depending on something from some external resource I guess, tiles / map data or something like that right?

Reply:
I think I will create a Level object that can contain all of these 16, 32, ... stuff from ldtk and it can also contain the map for pathfinding then.

Reply:
No, I change my mind... I think I will just have a constant for this for the time being and introduce Level if we will have more levels. I think Level handling would be the right solution, but I would not like to introduce further unnecessary complexity.

Reply:
Ah, I'm happy I found this comment again and I'm glad you changed your mind cos I was going to ask you not to do this

> I think I will create a Level object that can contain all of these 16, 32, ... stuff from ldtk

What advantage would this bring? Just a shorter name to call it by? These values are already provided by ldtk-go we just need to use them. For example, if the 32 here is the layer's grid size, that's found in [layer.GridSize](https://pkg.go.dev/github.com/solarlune/ldtkgo#Layer). A usage example from our own code: [961ae68/main.go#L133-L134](https://github.com/sinisterstuf/escort-mission/blob/961ae684345066f358bc63b99f0fc14172fbfa56/main.go#L133-L134). The 16 here I'm not sure what it represents (because it has no name) but I suppose it's GridSize/2 to be able to get the centre of it, but it may also be some other value from the map data, in which case I'm pretty sure we can find it from the ldtk-go-provided data 😉